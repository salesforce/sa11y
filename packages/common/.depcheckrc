# Ignore webdriverio - we need the types for `WebdriverIO.Browser`
ignores: ['webdriverio']
