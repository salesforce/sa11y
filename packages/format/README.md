# `format`

Format accessibility results from axe

## Usage

```
import axe from 'axe-core';
import { a11yResultsFormatter } from '@sa11y/format';

const results = await axe.run();
console.log(a11yResultsFormatter(results.violations));
```
