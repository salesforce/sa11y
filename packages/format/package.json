{
    "name": "@sa11y/format",
    "version": "4.0.0-alpha.0",
    "description": "Accessibility results re-formatter",
    "license": "BSD-3-Clause",
    "homepage": "https://github.com/salesforce/sa11y/tree/master/packages/format#readme",
    "repository": {
        "type": "git",
        "url": "https://github.com/salesforce/sa11y.git",
        "directory": "packages/format"
    },
    "keywords": [
        "accessibility",
        "automated testing",
        "axe",
        "results",
        "format",
        "template"
    ],
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "files": [
        "dist/**/*.js",
        "dist/**/*.d.ts*"
    ],
    "dependencies": {
        "@sa11y/common": "4.0.0-alpha.0",
        "@sa11y/preset-rules": "4.0.0-alpha.0",
        "axe-core": "4.7.0"
    },
    "devDependencies": {
        "@jest/globals": "28.1.3",
        "@sa11y/test-utils": "4.0.0-alpha.0"
    },
    "publishConfig": {
        "access": "public"
    },
    "engines": {
        "node": "^14 || ^16"
    }
}
