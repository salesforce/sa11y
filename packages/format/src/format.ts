/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { Result } from 'axe-core';

/**
 * Formatter defines the function signature to format accessibility violations found by axe
 */
export interface Formatter {
    (violations: Result[]): string;
}

/**
 * Format accessibility results from axe
 * @param violations - Result list from axe
 * */
// TODO: Add handlebars template for formatting
// TODO: Add support for different output formats console(colored), plain text, HTML, xUnit
export function a11yResultsFormatter(violations: Result[]): string {
    return violations
        .map((violation) => {
            return violation.nodes
                .map((node) => {
                    const selectors = node.target.join(', ');
                    return `${violation.help} (${violation.id}): ${selectors}\n  - More info: ${violation.helpUrl}`;
                })
                .join('\n\n');
        })
        .join('\n\n');
}
