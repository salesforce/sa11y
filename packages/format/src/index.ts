/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { A11yError, Options } from './format';
export { exceptionListFilter } from './filter';
export { A11yResult, A11yResults } from './result';
