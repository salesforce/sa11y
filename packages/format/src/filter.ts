/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import * as axe from 'axe-core';
import { AxeResults, ExceptionList } from '@sa11y/common';

/**
 * Filter a11y violations from axe based on given {@link ExceptionList}
 * @param violations - List of violations found with axe
 * @param exceptionList - {@link ExceptionList} of map of rule to corresponding css targets that needs to be filtered from a11y results
 */
export function exceptionListFilter(violations: AxeResults, exceptionList: ExceptionList = {}): AxeResults {
    const exceptionRules = Object.keys(exceptionList);
    if (exceptionRules.length === 0) return violations;

    const filteredViolations: AxeResults = [];

    for (const violation of violations) {
        if (!exceptionRules.includes(violation.id)) {
            filteredViolations.push(violation);
        } else {
            for (const result of violation.nodes) {
                const filteredResults = [];
                result.target.forEach((cssSelectorItem) => {
                    if (Array.isArray(cssSelectorItem)) {
                        cssSelectorItem.forEach((cssSelector) => {
                            if (!exceptionList[violation.id].includes(cssSelector)) {
                                filteredResults.push(cssSelector);
                            }
                        });
                    } else {
                        if (!exceptionList[violation.id].includes(cssSelectorItem)) {
                            filteredResults.push(cssSelectorItem);
                        }
                    }
                });
                if (filteredResults.length > 0) {
                    filteredViolations.push(violation);
                }
            }
        }
    }

    return filteredViolations;
}

/**
 * Filter a11y violations from axe based on given selectors filter keywords
 * @param violations - List of violations found with axe
 * @param selectorFilterKeywords - List of selector keywords to filter violations for
 */
export function exceptionListFilterSelectorKeywords(
    violations: AxeResults,
    selectorFilterKeywords: string[]
): AxeResults {
    const filteredViolations: AxeResults = [];
    for (const violation of violations) {
        const filteredNodes: axe.NodeResult[] = [];
        for (const node of violation.nodes) {
            const isSelectorFilterKeywordsExists = checkSelectorFilterKeyWordsExists(node, selectorFilterKeywords);
            if (!isSelectorFilterKeywordsExists) {
                filteredNodes.push(node);
            }
        }
        if (filteredNodes.length > 0) {
            violation.nodes = filteredNodes;
            filteredViolations.push(violation);
        }
    }
    return filteredViolations;
}

function checkSelectorFilterKeyWordsExists(node: axe.NodeResult, selectorFilterKeywords: string[]): boolean {
    const selectorAncestry = (node.ancestry?.flat(Infinity) ?? []) as string[];
    let isExists = false;
    selectorFilterKeywords.some((keyword) => {
        isExists = selectorAncestry.some((selector) => {
            const lastSelector = selector.split('>').pop();
            return lastSelector?.includes(keyword) ?? false;
        });
        return isExists;
    });
    return isExists;
}
