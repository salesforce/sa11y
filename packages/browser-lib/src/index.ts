/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import * as axe from 'axe-core';
import { exceptionListFilter, appendWcag } from '@sa11y/format';
import { defaultRuleset, registerCustomRules } from '@sa11y/preset-rules';
export { base, extended, full } from '@sa11y/preset-rules';
export const namespace = 'sa11y';

/**
 * Wrapper function to check accessibility to be invoked after the sa11y minified JS file
 * is injected into the browser
 * @param scope - Scope of the analysis as {@link A11yCheckableContext}
 * @param rules - preset sa11y rules defaulting to {@link base}
 * @param exceptionList - mapping of rule to css selectors to be filtered out using {@link exceptionListFilter}
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function checkAccessibility(
    scope = document,
    rules = defaultRuleset,
    exceptionList = {},
    addWcagInfo = false
) {
    // TODO (debug): adding type annotations to args, return type results in error:
    //  "[!] Error: Unexpected token" in both rollup-plugin-typescript2 and @rollup/plugin-typescript
    //  Compiling the index.ts file with tsc and using the dist/index.js file results
    //  in error when importing the "namespace" var. This would probably be easier to fix
    //  which could then result in getting rid of the rollup typescript plugins.

    // To register custom rules
    registerCustomRules();

    const results = await axe.run(scope || document, rules);
    const filteredResults = exceptionListFilter(results.violations, exceptionList);

    if (addWcagInfo) {
        appendWcag(filteredResults);
    }
    return JSON.stringify(filteredResults);
}
