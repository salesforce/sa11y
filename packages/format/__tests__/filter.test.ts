/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { exceptionListFilter, exceptionListFilterSelectorKeywords } from '../src';
import { AxeResults } from '@sa11y/common';
import { getViolations } from './format.test';
import { expect } from '@jest/globals';

const mockViolations = [
    {
        id: 'aria-allowed-attr',
        impact: 'serious',
        tags: ['cat.aria', 'wcag2a', 'wcag412'],
        description: "Ensures ARIA attributes are allowed for an element's role",
        help: 'Elements must only use allowed ARIA attributes',
        helpUrl: 'https://dequeuniversity.com/rules/axe/4.7/aria-allowed-attr?application=axeAPI',
        nodes: [
            {
                any: [],
                all: [],
                none: [
                    {
                        id: 'aria-prohibited-attr',
                        data: {
                            role: null,
                            nodeName: 'lightning-button-icon',
                            messageKey: 'noRoleSingular',
                            prohibited: ['aria-label'],
                        },
                        relatedNodes: [],
                        impact: 'serious',
                        message:
                            'aria-label attribute cannot be used on a lightning-button-icon with no valid role attribute.',
                    },
                ],
                impact: 'serious',
                html: '<lightning-button-icon lwc-2pdnejk934a="" class="slds-button slds-button_icon slds-button_icon-small slds-float_right slds-popover__close" aria-label="AgentWhisper.CloseDialog" title="AgentWhisper.CloseDialog"></lightning-button-icon>',
                target: ['lightning-button-icon'],
            },
        ],
    },
    {
        id: 'aria-dialog-name',
        impact: 'serious',
        tags: ['cat.aria', 'best-practice'],
        description: 'Ensures every ARIA dialog and alertdialog node has an accessible name',
        help: 'ARIA dialog and alertdialog nodes should have an accessible name',
        helpUrl: 'https://dequeuniversity.com/rules/axe/4.7/aria-dialog-name?application=axeAPI',
        nodes: [
            {
                any: [
                    {
                        id: 'aria-label',
                        data: null,
                        relatedNodes: [],
                        impact: 'serious',
                        message: 'aria-label attribute does not exist or is empty',
                    },
                    {
                        id: 'aria-labelledby',
                        data: null,
                        relatedNodes: [],
                        impact: 'serious',
                        message:
                            'aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty',
                    },
                    {
                        id: 'non-empty-title',
                        data: {
                            messageKey: 'noAttr',
                        },
                        relatedNodes: [],
                        impact: 'serious',
                        message: 'Element has no title attribute',
                    },
                ],
                all: [],
                none: [],
                impact: 'serious',
                html: '<section lwc-2pdnejk934a="" class="slds-popover popover-position-fixed slds-nubbin_top popover-container-bottom" aria-describedby="dialog-body-id-113-0" aria-labelledby="dialog-heading-id-5-0" role="dialog" style="--scrollOffset: 0px;">',
                target: ['section'],
            },
        ],
    },
];

let violations: AxeResults = [];
beforeAll(async () => {
    violations = await getViolations();
});

describe('a11y results filter', () => {
    it('should be setup with violations that can be reused across tests', () => {
        expect(violations).toMatchSnapshot();
    });

    it('should not filter results for default empty exception list', () => {
        expect(exceptionListFilter(violations)).toStrictEqual(violations);
    });

    it('should not filter results for empty violations list)', () => {
        expect(exceptionListFilter([])).toStrictEqual([]);
    });

    it.each([
        {},
        { nonExistingRule: ['foo', 'bar'] },
        { bypass: ['html5'] },
        { bypassFoo: ['html'] },
        { 'link-name': ['html'] },
        { 'document-title': ['a'], 'link-name': ['html'], 'bypass': ['a'] },
    ])('should not filter results for non-matching exception list %#', (exceptionList) => {
        expect(exceptionListFilter(violations, exceptionList)).toStrictEqual(violations);
    });

    it.each([
        { 'document-title': ['html'] },
        { 'link-name': ['a'] },
        { 'document-title': ['html'], 'link-name': ['a'] },
    ])('should filter results for matching exception list %#', (exceptionList) => {
        const filteredViolations = exceptionListFilter(violations, exceptionList);
        expect(filteredViolations).not.toStrictEqual(violations);
        const filteredRuleIDs = filteredViolations.map((violation) => violation.id);
        expect(filteredRuleIDs).not.toContain(Object.keys(exceptionList));
        expect(filteredViolations).toMatchSnapshot();
    });

    it('should filter results appropriately for exception list with both matching and non-matching items', () => {
        const validRule = 'document-title'; // Rule with valid css selector that matches violations
        const mixedExceptionList = { 'document-title': ['html'], 'link-name': ['html'], 'bypass': ['a'] };
        const ruleIDs = violations.map((violation) => violation.id);
        const filteredViolations = exceptionListFilter(violations, mixedExceptionList);
        const filteredRuleIDs = filteredViolations.map((violation) => violation.id);
        expect(filteredRuleIDs).not.toContain(validRule);
        expect(ruleIDs).toContain(validRule);
        expect(ruleIDs.filter((ruleID) => ruleID !== validRule)).toStrictEqual(filteredRuleIDs);
    });

    it('should filter results appropriately for exception list with multi-array result target', () => {
        const modifiedViolations: AxeResults = JSON.parse(JSON.stringify(violations)) as AxeResults;
        modifiedViolations.forEach((violation) => {
            violation.nodes.forEach((node) => {
                node.target = [node.target];
            });
        });
        const validRule = 'document-title'; // Rule with valid css selector that matches violations
        const mixedExceptionList = { 'document-title': ['html'], 'link-name': ['html'], 'bypass': ['a'] };
        const ruleIDs = modifiedViolations.map((violation) => violation.id);
        const filteredViolations = exceptionListFilter(modifiedViolations, mixedExceptionList);
        const filteredRuleIDs = filteredViolations.map((violation) => violation.id);
        expect(filteredRuleIDs).not.toContain(validRule);
        expect(ruleIDs).toContain(validRule);
        expect(ruleIDs.filter((ruleID) => ruleID !== validRule)).toStrictEqual(filteredRuleIDs);
    });

    it('should filter violations based on selector keywords', () => {
        // add ancestry keys
        mockViolations[0].nodes[0]['ancestry'] = [
            'html > body > agent-whisper-popover > section > lightning-button-icon:nth-child(1)',
        ];
        mockViolations[1].nodes[0]['ancestry'] = ['html > body > agent-whisper-popover > section'];

        const filteredViolations = exceptionListFilterSelectorKeywords(mockViolations as AxeResults, ['lightning-']);
        expect(filteredViolations).toHaveLength(1);
    });

    it('should not filter violations if no ancestry keys defined', () => {
        delete mockViolations[0].nodes[0]['ancestry'];
        delete mockViolations[1].nodes[0]['ancestry'];

        const filteredViolations = exceptionListFilterSelectorKeywords(mockViolations as AxeResults, ['lightning-']);
        expect(filteredViolations).toHaveLength(2);
    });
});
