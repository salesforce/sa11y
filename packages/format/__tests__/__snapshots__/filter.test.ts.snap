// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 0 1`] = `
"* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/link-name"
`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 0 2`] = `2`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 0 3`] = `
"2 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/link-name"
`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 1 1`] = `""`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 1 2`] = `0`;

exports[`a11y Results Formatter should format a11y issues as expected with default options for dom 1 3`] = `
"0 Accessibility issues found
 "
`;

exports[`a11y Results Formatter should format using specified options: 0 1`] = `"[{\\"id\\":\\"document-title\\",\\"impact\\":\\"serious\\",\\"tags\\":[\\"cat.text-alternatives\\",\\"wcag2a\\",\\"wcag242\\",\\"ACT\\",\\"TTv5\\",\\"TT12.a\\"],\\"description\\":\\"Ensures each HTML document contains a non-empty <title> element\\",\\"help\\":\\"Documents must have <title> element to aid in navigation\\",\\"helpUrl\\":\\"https://dequeuniversity.com/rules/axe/4.7/document-title?application=axeAPI\\",\\"nodes\\":[{\\"any\\":[{\\"id\\":\\"doc-has-title\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"Document does not have a non-empty <title> element\\"}],\\"all\\":[],\\"none\\":[],\\"impact\\":\\"serious\\",\\"html\\":\\"<html lang=\\\\\\"en\\\\\\"><head><style>* { pointer-events: all }</style></head><body>\\\\n\\\\n<div id=\\\\\\"dom-with-issues\\\\\\">\\\\n    <a href=\\\\\\"#\\\\\\"></a>\\\\n</div>\\\\n\\\\n\\\\n</body></html>\\",\\"target\\":[\\"html\\"],\\"failureSummary\\":\\"Fix any of the following:\\\\n  Document does not have a non-empty <title> element\\"}]},{\\"id\\":\\"link-name\\",\\"impact\\":\\"serious\\",\\"tags\\":[\\"cat.name-role-value\\",\\"wcag2a\\",\\"wcag412\\",\\"wcag244\\",\\"section508\\",\\"section508.22.a\\",\\"ACT\\",\\"TTv5\\",\\"TT6.a\\"],\\"description\\":\\"Ensures links have discernible text\\",\\"help\\":\\"Links must have discernible text\\",\\"helpUrl\\":\\"https://dequeuniversity.com/rules/axe/4.7/link-name?application=axeAPI\\",\\"nodes\\":[{\\"any\\":[{\\"id\\":\\"has-visible-text\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"minor\\",\\"message\\":\\"Element does not have text that is visible to screen readers\\"},{\\"id\\":\\"aria-label\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"aria-label attribute does not exist or is empty\\"},{\\"id\\":\\"aria-labelledby\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty\\"},{\\"id\\":\\"non-empty-title\\",\\"data\\":{\\"messageKey\\":\\"noAttr\\"},\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"Element has no title attribute\\"}],\\"all\\":[],\\"none\\":[{\\"id\\":\\"focusable-no-name\\",\\"data\\":null,\\"relatedNodes\\":[],\\"impact\\":\\"serious\\",\\"message\\":\\"Element is in tab order and does not have accessible text\\"}],\\"impact\\":\\"serious\\",\\"html\\":\\"<a href=\\\\\\"#\\\\\\"></a>\\",\\"target\\":[\\"a\\"],\\"failureSummary\\":\\"Fix all of the following:\\\\n  Element is in tab order and does not have accessible text\\\\n\\\\nFix any of the following:\\\\n  Element does not have text that is visible to screen readers\\\\n  aria-label attribute does not exist or is empty\\\\n  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty\\\\n  Element has no title attribute\\"}]}]"`;

exports[`a11y Results Formatter should format using specified options: 1 1`] = `
"\\"* (document-title) Documents must have <title> element to aid in navigation: html\\"
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/document-title

\\"* (link-name) Links must have discernible text: a\\"
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/link-name"
`;

exports[`a11y Results Formatter should format using specified options: 2 1`] = `
"* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/link-name"
`;

exports[`a11y Results Formatter should format using specified options: 3 1`] = `
"* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/link-name"
`;

exports[`a11y Results Formatter should format using specified options: 4 1`] = `
"* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/link-name"
`;

exports[`a11y Results Formatter should not throw error for repeated violations with consolidation 1`] = `
"2 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/link-name"
`;

exports[`a11y Results Formatter should not throw error for repeated violations with consolidation 2`] = `
"2 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/link-name"
`;

exports[`a11y Results Formatter should throw error when violations are present 1`] = `
"2 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.7/link-name"
`;

exports[`a11y results filter should be setup with violations that can be reused across tests 1`] = `
Array [
  Object {
    "description": "Ensures each HTML document contains a non-empty <title> element",
    "help": "Documents must have <title> element to aid in navigation",
    "helpUrl": "https://dequeuniversity.com/rules/axe/4.7/document-title?application=axeAPI",
    "id": "document-title",
    "impact": "serious",
    "nodes": Array [
      Object {
        "all": Array [],
        "any": Array [
          CheckResult {
            "data": null,
            "id": "doc-has-title",
            "impact": "serious",
            "message": "Document does not have a non-empty <title> element",
            "relatedNodes": Array [],
          },
        ],
        "failureSummary": "Fix any of the following:
  Document does not have a non-empty <title> element",
        "html": "<html><head></head><body>

<div id=\\"dom-with-issues\\">
    <a href=\\"#\\"></a>
</div>


</body></html>",
        "impact": "serious",
        "none": Array [],
        "target": Array [
          "html",
        ],
      },
    ],
    "tags": Array [
      "cat.text-alternatives",
      "wcag2a",
      "wcag242",
      "ACT",
      "TTv5",
      "TT12.a",
    ],
  },
  Object {
    "description": "Ensures every HTML document has a lang attribute",
    "help": "<html> element must have a lang attribute",
    "helpUrl": "https://dequeuniversity.com/rules/axe/4.7/html-has-lang?application=axeAPI",
    "id": "html-has-lang",
    "impact": "serious",
    "nodes": Array [
      Object {
        "all": Array [],
        "any": Array [
          CheckResult {
            "data": Object {
              "messageKey": "noLang",
            },
            "id": "has-lang",
            "impact": "serious",
            "message": "The <html> element does not have a lang attribute",
            "relatedNodes": Array [],
          },
        ],
        "failureSummary": "Fix any of the following:
  The <html> element does not have a lang attribute",
        "html": "<html><head></head><body>

<div id=\\"dom-with-issues\\">
    <a href=\\"#\\"></a>
</div>


</body></html>",
        "impact": "serious",
        "none": Array [],
        "target": Array [
          "html",
        ],
      },
    ],
    "tags": Array [
      "cat.language",
      "wcag2a",
      "wcag311",
      "ACT",
      "TTv5",
      "TT11.a",
    ],
  },
  Object {
    "description": "Ensures links have discernible text",
    "help": "Links must have discernible text",
    "helpUrl": "https://dequeuniversity.com/rules/axe/4.7/link-name?application=axeAPI",
    "id": "link-name",
    "impact": "serious",
    "nodes": Array [
      Object {
        "all": Array [],
        "any": Array [
          CheckResult {
            "data": null,
            "id": "has-visible-text",
            "impact": "minor",
            "message": "Element does not have text that is visible to screen readers",
            "relatedNodes": Array [],
          },
          CheckResult {
            "data": null,
            "id": "aria-label",
            "impact": "serious",
            "message": "aria-label attribute does not exist or is empty",
            "relatedNodes": Array [],
          },
          CheckResult {
            "data": null,
            "id": "aria-labelledby",
            "impact": "serious",
            "message": "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty",
            "relatedNodes": Array [],
          },
          CheckResult {
            "data": Object {
              "messageKey": "noAttr",
            },
            "id": "non-empty-title",
            "impact": "serious",
            "message": "Element has no title attribute",
            "relatedNodes": Array [],
          },
        ],
        "failureSummary": "Fix all of the following:
  Element is in tab order and does not have accessible text

Fix any of the following:
  Element does not have text that is visible to screen readers
  aria-label attribute does not exist or is empty
  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
  Element has no title attribute",
        "html": "<a href=\\"#\\"></a>",
        "impact": "serious",
        "none": Array [
          CheckResult {
            "data": null,
            "id": "focusable-no-name",
            "impact": "serious",
            "message": "Element is in tab order and does not have accessible text",
            "relatedNodes": Array [],
          },
        ],
        "target": Array [
          "a",
        ],
      },
    ],
    "tags": Array [
      "cat.name-role-value",
      "wcag2a",
      "wcag412",
      "wcag244",
      "section508",
      "section508.22.a",
      "ACT",
      "TTv5",
      "TT6.a",
    ],
  },
]
`;

exports[`a11y results filter should filter results for matching exception list 0 1`] = `
Array [
  Object {
    "description": "Ensures every HTML document has a lang attribute",
    "help": "<html> element must have a lang attribute",
    "helpUrl": "https://dequeuniversity.com/rules/axe/4.7/html-has-lang?application=axeAPI",
    "id": "html-has-lang",
    "impact": "serious",
    "nodes": Array [
      Object {
        "all": Array [],
        "any": Array [
          CheckResult {
            "data": Object {
              "messageKey": "noLang",
            },
            "id": "has-lang",
            "impact": "serious",
            "message": "The <html> element does not have a lang attribute",
            "relatedNodes": Array [],
          },
        ],
        "failureSummary": "Fix any of the following:
  The <html> element does not have a lang attribute",
        "html": "<html><head></head><body>

<div id=\\"dom-with-issues\\">
    <a href=\\"#\\"></a>
</div>


</body></html>",
        "impact": "serious",
        "none": Array [],
        "target": Array [
          "html",
        ],
      },
    ],
    "tags": Array [
      "cat.language",
      "wcag2a",
      "wcag311",
      "ACT",
      "TTv5",
      "TT11.a",
    ],
  },
  Object {
    "description": "Ensures links have discernible text",
    "help": "Links must have discernible text",
    "helpUrl": "https://dequeuniversity.com/rules/axe/4.7/link-name?application=axeAPI",
    "id": "link-name",
    "impact": "serious",
    "nodes": Array [
      Object {
        "all": Array [],
        "any": Array [
          CheckResult {
            "data": null,
            "id": "has-visible-text",
            "impact": "minor",
            "message": "Element does not have text that is visible to screen readers",
            "relatedNodes": Array [],
          },
          CheckResult {
            "data": null,
            "id": "aria-label",
            "impact": "serious",
            "message": "aria-label attribute does not exist or is empty",
            "relatedNodes": Array [],
          },
          CheckResult {
            "data": null,
            "id": "aria-labelledby",
            "impact": "serious",
            "message": "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty",
            "relatedNodes": Array [],
          },
          CheckResult {
            "data": Object {
              "messageKey": "noAttr",
            },
            "id": "non-empty-title",
            "impact": "serious",
            "message": "Element has no title attribute",
            "relatedNodes": Array [],
          },
        ],
        "failureSummary": "Fix all of the following:
  Element is in tab order and does not have accessible text

Fix any of the following:
  Element does not have text that is visible to screen readers
  aria-label attribute does not exist or is empty
  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
  Element has no title attribute",
        "html": "<a href=\\"#\\"></a>",
        "impact": "serious",
        "none": Array [
          CheckResult {
            "data": null,
            "id": "focusable-no-name",
            "impact": "serious",
            "message": "Element is in tab order and does not have accessible text",
            "relatedNodes": Array [],
          },
        ],
        "target": Array [
          "a",
        ],
      },
    ],
    "tags": Array [
      "cat.name-role-value",
      "wcag2a",
      "wcag412",
      "wcag244",
      "section508",
      "section508.22.a",
      "ACT",
      "TTv5",
      "TT6.a",
    ],
  },
]
`;

exports[`a11y results filter should filter results for matching exception list 1 1`] = `
Array [
  Object {
    "description": "Ensures each HTML document contains a non-empty <title> element",
    "help": "Documents must have <title> element to aid in navigation",
    "helpUrl": "https://dequeuniversity.com/rules/axe/4.7/document-title?application=axeAPI",
    "id": "document-title",
    "impact": "serious",
    "nodes": Array [
      Object {
        "all": Array [],
        "any": Array [
          CheckResult {
            "data": null,
            "id": "doc-has-title",
            "impact": "serious",
            "message": "Document does not have a non-empty <title> element",
            "relatedNodes": Array [],
          },
        ],
        "failureSummary": "Fix any of the following:
  Document does not have a non-empty <title> element",
        "html": "<html><head></head><body>

<div id=\\"dom-with-issues\\">
    <a href=\\"#\\"></a>
</div>


</body></html>",
        "impact": "serious",
        "none": Array [],
        "target": Array [
          "html",
        ],
      },
    ],
    "tags": Array [
      "cat.text-alternatives",
      "wcag2a",
      "wcag242",
      "ACT",
      "TTv5",
      "TT12.a",
    ],
  },
  Object {
    "description": "Ensures every HTML document has a lang attribute",
    "help": "<html> element must have a lang attribute",
    "helpUrl": "https://dequeuniversity.com/rules/axe/4.7/html-has-lang?application=axeAPI",
    "id": "html-has-lang",
    "impact": "serious",
    "nodes": Array [
      Object {
        "all": Array [],
        "any": Array [
          CheckResult {
            "data": Object {
              "messageKey": "noLang",
            },
            "id": "has-lang",
            "impact": "serious",
            "message": "The <html> element does not have a lang attribute",
            "relatedNodes": Array [],
          },
        ],
        "failureSummary": "Fix any of the following:
  The <html> element does not have a lang attribute",
        "html": "<html><head></head><body>

<div id=\\"dom-with-issues\\">
    <a href=\\"#\\"></a>
</div>


</body></html>",
        "impact": "serious",
        "none": Array [],
        "target": Array [
          "html",
        ],
      },
    ],
    "tags": Array [
      "cat.language",
      "wcag2a",
      "wcag311",
      "ACT",
      "TTv5",
      "TT11.a",
    ],
  },
]
`;

exports[`a11y results filter should filter results for matching exception list 2 1`] = `
Array [
  Object {
    "description": "Ensures every HTML document has a lang attribute",
    "help": "<html> element must have a lang attribute",
    "helpUrl": "https://dequeuniversity.com/rules/axe/4.7/html-has-lang?application=axeAPI",
    "id": "html-has-lang",
    "impact": "serious",
    "nodes": Array [
      Object {
        "all": Array [],
        "any": Array [
          CheckResult {
            "data": Object {
              "messageKey": "noLang",
            },
            "id": "has-lang",
            "impact": "serious",
            "message": "The <html> element does not have a lang attribute",
            "relatedNodes": Array [],
          },
        ],
        "failureSummary": "Fix any of the following:
  The <html> element does not have a lang attribute",
        "html": "<html><head></head><body>

<div id=\\"dom-with-issues\\">
    <a href=\\"#\\"></a>
</div>


</body></html>",
        "impact": "serious",
        "none": Array [],
        "target": Array [
          "html",
        ],
      },
    ],
    "tags": Array [
      "cat.language",
      "wcag2a",
      "wcag311",
      "ACT",
      "TTv5",
      "TT11.a",
    ],
  },
]
`;
